File: oaiVoiceModeChat/OpenAIModels.swift

struct ConversationItem: Decodable {
    let id: String?
    let title: String?
    let create_time: String
    let update_time: String
    let mapping: String?
    let current_node: String?
    let conversation_template_id: String?
    let gizmo_id: String?
    let is_archived: Bool
    let is_starred: Bool?
    let is_unread: Bool?
    let workspace_id: String?
    let async_status: String?
    let safe_urls: [String]?
    let conversation_origin: String?
    let snippet: String?
}

struct ConversationHistory: Decodable {
    let items: [ConversationItem]
}
struct Message: Decodable {
    let title: String?
    let create_time: Double?
    let update_time: Double?
    let mapping: [String: MessageNode]?
}

struct MessageNode: Decodable {
    let id: String
    let message: MessageContent?
    let parent: String?
    let children: [String]
}

struct MessageContent: Decodable {
    let id: String?
    let author: Author?
    let create_time: Double?
    let update_time: Double?
    let content: InnerContent?
    let status: String?
    let end_turn: Bool?
    let weight: Double?
    let metadata: MessageMetadata?
    let recipient: String?
    let channel: String?
}

struct Author: Decodable {
    let role: String
    let name: String?
    let metadata: [String: String]
}

struct InnerContent: Decodable {
    let content_type: String
    let parts: [ContentPartEnum?]?
}

enum ContentPartEnum: Decodable {
    case contentPart(ContentPart)
    case string(String)

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let string = try? container.decode(String.self) {
            self = .string(string)
        } else if let contentPart = try? container.decode(ContentPart.self) {
            self = .contentPart(contentPart)
        } else {
            throw DecodingError.typeMismatch(
                ContentPartEnum.self,
                DecodingError.Context(
                    codingPath: decoder.codingPath,
                    debugDescription: "Expected String or ContentPart"
                )
            )
        }
    }
}

struct ContentPart: Decodable {
    let content_type: String?
    let text: String?
    let direction: String?
    let decoding_id: String?
    let expiry_datetime: String?
    let frames_asset_pointers: [String]?
    let video_container_asset_pointer: String?
    let audio_asset_pointer: AudioAssetPointer?
    let audio_start_timestamp: Double?
}

struct AudioAssetPointer: Decodable {
    let expiry_datetime: String?
    let content_type: String
    let asset_pointer: String
    let size_bytes: Int
    let format: String
    let metadata: AudioMetadata
}

struct AudioMetadata: Decodable {
    let start_timestamp: Double?
    let end_timestamp: Double?
    let pretokenized_vq: String?
    let interruptions: String?
    let original_audio_source: String?
    let transcription: String?
    let start: Double?
    let end: Double?
}

struct MessageMetadata: Decodable {
    let voice_mode_message: Bool?
    let request_id: String?
    let message_source: String?
    let timestamp_: String?
    let message_type: String?
    let real_time_audio_has_video: Bool?
    let is_visually_hidden_from_conversation: Bool?
    let citations: [String]?
    let content_references: [String]?
    let is_complete: Bool?
    let parent_id: String?
    let model_switcher_deny: [ModelSwitcherDeny]?
}

struct ModelSwitcherDeny: Decodable {
    let slug: String
    let context: String
    let reason: String
    let description: String
}

File: oaiVoiceModeChat/openai_models.py
from pydantic import BaseModel
from typing import List, Optional, Union

class ConversationItem(BaseModel):
    id: Optional[str]
    title: Optional[str]
    create_time: str
    update_time: str
    mapping: Optional[str]
    current_node: Optional[str]
    conversation_template_id: Optional[str]
    gizmo_id: Optional[str]
    is_archived: bool
    is_starred: Optional[bool]
    is_unread: Optional[bool]
    workspace_id: Optional[str]
    async_status: Optional[str]
    safe_urls: Optional[List[str]]
    conversation_origin: Optional[str]
    snippet: Optional[str]

class ConversationHistory(BaseModel):
    items: List[ConversationItem]

class Message(BaseModel):
    title: Optional[str]
    create_time: Optional[float]
    update_time: Optional[float]
    mapping: Optional[dict]

class MessageNode(BaseModel):
    id: str
    message: Optional['MessageContent']
    parent: Optional[str]
    children: List[str]

class MessageContent(BaseModel):
    id: Optional[str]
    author: Optional['Author']
    create_time: Optional[float]
    update_time: Optional[float]
    content: Optional['InnerContent']
    status: Optional[str]
    end_turn: Optional[bool]
    weight: Optional[float]
    metadata: Optional['MessageMetadata']
    recipient: Optional[str]
    channel: Optional[str]

class Author(BaseModel):
    role: str
    name: Optional[str]
    metadata: dict

class InnerContent(BaseModel):
    content_type: str
    parts: Optional[List[Union[str, 'ContentPart']]]

class ContentPart(BaseModel):
    content_type: Optional[str]
    text: Optional[str]
    direction: Optional[str]
    decoding_id: Optional[str]
    expiry_datetime: Optional[str]
    frames_asset_pointers: Optional[List[str]]
    video_container_asset_pointer: Optional[str]
    audio_asset_pointer: Optional['AudioAssetPointer']
    audio_start_timestamp: Optional[float]

class AudioAssetPointer(BaseModel):
    expiry_datetime: Optional[str]
    content_type: str
    asset_pointer: str
    size_bytes: int
    format: str
    metadata: 'AudioMetadata'

class AudioMetadata(BaseModel):
    start_timestamp: Optional[float]
    end_timestamp: Optional[float]
    pretokenized_vq: Optional[str]
    interruptions: Optional[str]
    original_audio_source: Optional[str]
    transcription: Optional[str]
    start: Optional[float]
    end: Optional[float]

class MessageMetadata(BaseModel):
    voice_mode_message: Optional[bool]
    request_id: Optional[str]
    message_source: Optional[str]
    timestamp_: Optional[str]
    message_type: Optional[str]
    real_time_audio_has_video: Optional[bool]
    is_visually_hidden_from_conversation: Optional[bool]
    citations: Optional[List[str]]
    content_references: Optional[List[str]]
    is_complete: Optional[bool]
    parent_id: Optional[str]
    model_switcher_deny: Optional[List['ModelSwitcherDeny']]

class ModelSwitcherDeny(BaseModel):
    slug: str
    context: str
    reason: str
    description: str
